### PR TITLE
Core: allow authentication to be configured on API endpoints

### PR DESCRIPTION
**Pull request by reinkrul**
Closes https://github.com/nuts-foundation/nuts-node/issues/540

Changes:
- Moves Echo handandling middleware configuration to a separate engine, because it became more and more which didn't belong in `core`.
- Added `http gen-token [user] [validity]` command to generate an API auth token
- Added `--token` and `--token-file` to client config for supplying the token to be used, or a file containing the token. If these are not supplied, the token is read from the user's home dir from a file named `.nuts-client.cfg`. Because of its name we can use it in the future to allow more elaborate config (e.g. mapping server->token, timeout properties etc).
- Added middleware to authenticate auth token
- Altered middleware to only apply if it wasn't already applied from the default HTTP interface. E.g., if one configures `/internal` without binding it on a separate  HTTP interface. It is then exposed on the default interface, but with its own middleware configuration (e.g. authentication). However, this could lead to middleware of the default interface being applied first, then the same middleware for the specific configuration. This was the case before, but didn't cause any problems because it was only the CORS middleware. With authentication (and later on request logging) it becomes more of a problem. This is now fixed by skipping the middleware if it's a configured path, which has its own middleware.
- Added optional JWT auth to all OpenAPI specs
- Moved `EchoCreator` which creates Echo servers for `MultiEcho` to being an argument of `MultiEcho.Bind()`, which solves many dependency issues.
- Moved middleware config out of the Echo creator function to the `Configure()` phase of the HTTP engine
- Introduced interfaces and adapter structs to have authentication headers apply to `HTTPClient` instances without having to write lots of boilerplate for each.
- Removed config properties from `HTTPClient`s and just have it embed `ClientConfig` for better maintainability (e.g. if we introduce a new property).

TODO:
- [x] Documentation
- [x] Check all HTTPClient implementations whether they really use `MustCreateHTTPClient()`
- [x] Rename "Nuts Node APIs" to "API Reference"
- [x] Introduce an "Integrating" page "API Authentication" describing how to configure and use it, refer to it from production config page
- [x] Move echo/integration test to core
- [x] Add test for skipper /metrics /status
- [x] Disallow subpaths in alt binds
- [x] Sanitize alt bind paths before passing them into MultiEcho

### CODE REVIEW

File: http/engine.go [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+		echoServer.TLSServer.TLSConfig = serverTLSConfig
+		echoServer.TLSServer.Addr = cfg.Address
+		startFn = func(address string) error {
+			return echoServer.StartServer(echoServer.TLSServer)
+		}
+	default:
+		fallthrough
+	case TLSDisabledMode:
+		startFn = echoServer.Start
+	}
```
Comment (reinkrul) [inline http/engine.go:134]:
```suggestion
	case TLSMode(""):
		fallthrough
	case TLSDisabledMode:
		startFn = echoServer.Start
	default:
		return nil, fmt.Errorf("invalid TLS mode: %s", cfg.TLSMode)
	}
```

File: http/ratelimiter.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+import (
+	"github.com/labstack/echo/v4"
+	"github.com/labstack/echo/v4/middleware"
+	"golang.org/x/time/rate"
+	"time"
+)
+
+// InternalRateLimiterStore uses a simple TokenBucket for limiting the amount of internal requests.
+// It should only be used for internal paths since it does not register the rate limit per caller.
+type InternalRateLimiterStore struct {
```
Comment (reinkrul) [inline http/ratelimiter.go:12]: unexport

**Review by gerardsn — CHANGES_REQUESTED** [review]
Nice work.
It is a large PR, so many comments/questions.

Still needed
- [ ] Some documentation on the lifecycle or expected usage of the tokens. (Ignore this point if this is just a gap in my knowledge.)
- [ ] A lot of code and therefore tests have moved around and some test appear to be deleted. Coverage seems to be missing some things as well, but not sure if this is due to this PR. A double check on this would be great.

File: core/http_client.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+package core
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+)
+
+func TestResponseCode(expectedStatusCode int, response *http.Response) error {
```
[the remark this answers is not in the document]
Comment (reinkrul) [reply core/http_client.go:28]: out of scope, but will do

File: docs/_static/auth/v1.yaml [hunk 4]
```diff
@@ -671,3 +671,11 @@ components:
           type: string
           description: End-User's preferred e-mail address. Should be a personal email and can be used to uniquely identify a user. Just like the email used for an account.
           example: w.debruijn@example.org
+  securitySchemes:
```
[the remark this answers is not in the document]
Comment (reinkrul) [reply docs/_static/auth/v1.yaml:674]: Unfortunately not

File: core/client_config.go [hunk 5]
```diff
@@ -32,12 +35,18 @@ const defaultClientTimeout = 10 * time.Second
...
+var userHomeDirFn = os.UserHomeDir
 
 // ClientConfig has CLI client settings.
 type ClientConfig struct {
 	Address   string        `koanf:"address"`
 	Verbosity string        `koanf:"verbosity"`
 	Timeout   time.Duration `koanf:"timeout"`
+	Token     string        `koanf:"token"`
+	TokenFile string        `koanf:"token-file"`
 }
```
[the remark this answers is not in the document]
Comment (reinkrul) [reply core/client_config.go:50]: Exactly.

File: docs/pages/deployment/cli-reference.rst [hunk 6]
```diff
@@ -169,9 +173,77 @@ Deletes a service from a DID document.
...
-      --timeout duration   Client time-out when performing remote operations, such as '500ms' or '10s'. Refer to Golang's 'time.Duration' syntax for a more elaborate description of the syntax. (default 10s)
-      --verbosity string   Log level (trace, debug, info, warn, error) (default "info")
+      --address string      Address of the remote node. Must contain at least host and port, URL scheme may be omitted. In that case it 'http://' is prepended. (default "localhost:1323")
+      --timeout duration    Client time-out when performing remote operations, such as '500ms' or '10s'. Refer to Golang's 'time.Duration' syntax for a more elaborate description of the syntax. (default 10s)
+      --token string        Token to be used for authenticating on the remote node. Takes precedence over 'token-file'.
+      --token-file string   File from which the authentication token will be read. If not specified it will try to read the token from the '.nuts-client.cfg' file in the user's home dir.
+      --verbosity string    Log level (trace, debug, info, warn, error) (default "info")
+
+nuts http gen-token
```
[the remark this answers is not in the document]
Comment (reinkrul) [reply docs/pages/deployment/cli-reference.rst:182]: https://github.com/nuts-foundation/nuts-node/issues/1377

File: echo/echo_integration_test.go [hunk 7]
```diff
@@ -22,6 +22,7 @@ package echo
 import (
```
[the remark this answers is not in the document]
Comment (reinkrul) [reply echo/echo_integration_test.go:22]: to core

File: main.go [hunk 8]
```diff
@@ -31,9 +31,12 @@ import (
...
 	// Listen for interrupt signals (CTRL/CMD+C, OS instructing the process to stop) to cancel context.
 	ctx := context.Background()
-	ctx, _ = signal.NotifyContext(ctx, os.Interrupt, syscall.SIGTERM)
+	ctx, cancelNotify := signal.NotifyContext(ctx, os.Interrupt, syscall.SIGTERM)
+	defer cancelNotify()
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
 
-	err := cmd.Execute(ctx, cmd.CreateSystem())
+	err := cmd.Execute(ctx, cmd.CreateSystem(cancel))
```
[the remark this answers is not in the document]
Comment (reinkrul) [reply main.go:39]: Probably nothing

File: core/client_config.go [hunk 9]
```diff
@@ -32,12 +35,18 @@ const defaultClientTimeout = 10 * time.Second
 const clientTimeoutFlag = "timeout"
 const clientAddressFlag = "address"
 const defaultAddress = "localhost" + (":1323")
+const clientConfigFileName = ".nuts-client.cfg"
```
[the remark this answers is not in the document]
Comment (reinkrul) [reply core/client_config.go:38]: Will document it in new "API Authentication" page

File: core/client_config.go [hunk 10]
```diff
@@ -81,11 +90,41 @@ func (cfg ClientConfig) GetAddress() string {
...
+	mustExist := true
+	if len(tokenFile) == 0 {
+		// token-file not set, try to read from user home dir
+		mustExist = false
+		dir, err := userHomeDirFn()
+		if err != nil {
+			return "", fmt.Errorf("unable to read auth token from file: %w", err)
+		}
+		tokenFile = path.Join(dir, clientConfigFileName)
+	}
```
[the remark this answers is not in the document]
Comment (reinkrul) [reply core/client_config.go:109]: Will document it to make it less magic

File: core/client_config.go [hunk 11]
```diff
@@ -32,12 +35,18 @@ const defaultClientTimeout = 10 * time.Second
 const clientTimeoutFlag = "timeout"
 const clientAddressFlag = "address"
 const defaultAddress = "localhost" + (":1323")
+const clientConfigFileName = ".nuts-client.cfg"
+
+// userHomeDirFn is settable for testing purposes
+var userHomeDirFn = os.UserHomeDir
```
[the remark this answers is not in the document]
Comment (reinkrul) [reply core/client_config.go:41]: Discussed

File: http/engine.go [hunk 12]
```diff
@@ -0,0 +1,295 @@
...
+	return requestURI == path || strings.HasPrefix(requestURI, path)
+}
+
+func (h Engine) applyGlobalMiddleware(echoServer core.EchoRouter, serverConfig core.ServerConfig) {
+	// Use middleware to decode URL encoded path parameters like did%3Anuts%3A123 -> did:nuts:123
+	echoServer.Use(decodeURIPath)
+
+	skipper := func(c echo.Context) bool {
+		return !matchesPath(c.Request().RequestURI, "/metrics") && !matchesPath(c.Request().RequestURI, "/status")
+	}
```
[the remark this answers is not in the document]
Comment (reinkrul) [reply http/engine.go:238]: will add test

File: core/client_config.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (gerardsn) [reply core/client_config.go:38]: still needs to be added to documentation

File: http/engine.go [hunk 13]
```diff
@@ -0,0 +1,313 @@
...
+			},
+			http.MethodPut: {
+				"/internal/vdr/v1/did/:did",                // updating DIDs
+				"/internal/didman/v1/did/:did/contactinfo", // updating contactinfo in DID
+			}}, 24*time.Hour, 3000, 30),
+		)
+	}
+}
+
+func (h Engine) applyBindMiddleware(echoServer EchoServer, path string, excludePaths []string, serverConfig core.ServerConfig, cfg InterfaceConfig) error {
```
Comment (gerardsn) [inline http/engine.go:259]:
```yaml
http:
  default:
    address: :1323
  alt:
    internal: # auth config is NOT accepted when address is not specified. /internal does not require a token with this config
      auth:
        type: token
    n2n: # does require token when address is specified
      address: :1323
      auth:
        type: token
```
The config above does not require a token when `http.alt.internal.auth` is set but `http.alt.internal.address` is not. If this is expected behavior it should be clearly documented.

**Review by gerardsn — CHANGES_REQUESTED** [review]
Couple of things left

**Pull request merged**